File: agent-driven-architecture-main/examples/kotlin/spine/src/main/kotlin/adr/spine/boundary/Gate.kt
// ── spine/boundary/gate — the irreversibility gate, PRE-FOLD (G1/G6) ───────
// The shipped reference branched on `r.by`, an Actor the TOOL copied into its own
// payload, while the log recorded a different Actor the boundary stamped after the
// fold had already run. Two unreconciled values, and the stamp was causally
// INCAPABLE of gating.
//
// Here there is exactly one value, it is created AFTER the tool has returned, and
// the check runs BEFORE the fold — which is where 14.3 always said it goes.
//
// What it compares (14.3 implemented, not paraphrased): the CONFIRMING authority
// against the REQUESTING authority recorded in State. "A different actor than the
// one that issued the Request" is not implemented by `by == Human` — that
// implements "a human". It is implemented by comparing principals.
//
// A refusal is COMMITTED as ToolResult.Refused, so it re-folds without re-running
// the authorization check (G9). G6's "captured as an ordered G9 fixture" and its per-tenant budget's
// "verdict captured as an ordered G9 fixture" are the SAME mechanism, not two.

package adr.spine.boundary

import adr.contract.ToolResult
import adr.spine.ports.ConfirmPolicy
import adr.spine.pure.Gating
import adr.spine.pure.Registry
import adr.spine.pure.Signature

/**
 * The gate itself, over the registry and the confirm policy it is built with. The
 * per-step values — the result, the signature, the state — are the arguments.
 *
 * Constructed by the Boundary, never injected into it. `policy` is the seam a test is
 * meant to vary; the gate AROUND it is the invariant, and a gate that could be bound
 * from outside is a gate that could be bypassed from outside.
 */
class IrreversibilityGate<S>(
    private val registry: Registry<S>,
    private val policy: ConfirmPolicy,
) {
    fun check(result: ToolResult, sig: Signature, state: S): ToolResult {
        // An unresolvable action never had a verb to gate; it is already a decision.
        if (result is ToolResult.Unhandled) return result

        val verb = registry[result.tool]
            ?: return ToolResult.Unhandled(result.tool, "no registered verb")

        return when (val gating = verb.gating(state, result)) {
            Gating.Ungated -> result
            is Gating.NeedsConfirmation -> when {
                gating.requestedBy == null ->
                    ToolResult.Refused(result.tool, "no pending request")

                gating.requestedBy == sig.authority ->
                    ToolResult.Refused(
                        result.tool,
                        "self-confirm: the confirming authority is the requesting authority",
                    )

                !policy.mayConfirm(sig, result, gating.requestedBy) ->
                    ToolResult.Refused(result.tool, "authority may not confirm this action")

                else -> result
            }
        }
    }
}

File: agent-driven-architecture-main/examples/typescript/src/spine/boundary/gate.ts
// ── spine/boundary/gate — the irreversibility gate (G1, G6; 14.3) ──────────
// PRE-FOLD, AT THE BOUNDARY, KEYED ON AUTHORITY.
//
// The shipped reference branched inside the fold on an Actor the TOOL had
// copied into its own payload, while the boundary stamped a different Actor one
// line later — two unreconciled values, and the stamp was causally incapable of
// gating. Here there is only ONE value, `sig`, it is created after every tool
// has returned, and the gate reads it before the fold sees anything.
//
// 14.3 requires "a different ACTOR than the one that issued the Request".
// `by == Human` does not implement that — it implements "a human". What is
// implemented below is `sig.authority != the authority recorded in State when
// the Request was raised`: a different PRINCIPAL than the one that asked. That
// is what the sentence says, and it is what an unattended deployment needs:
//
//   confirmer                        sig.by   sig.authority     verdict
//   the agent re-confirming its own   Agent    agent-run-7f      REFUSED (self-confirm)
//   a policy tier approving by rules  Agent    policy-tier-v3    granted
//   a second-agent reviewer           Agent    reviewer-a2       granted
//   a human host                      Human    host:marcos       granted
//   a deferred approval queue         Agent    approval-queue    granted
//   the agent confirming a SPINE-
//     requested drain seal            Agent    agent-run-7f      granted
//
// The confirming Command still stamps its Actor TRUTHFULLY; the Authority is
// the field that differs. No second bus (5.2 holds); recall confers no
// authority (11.3 holds); and no row above adds an Actor variant — the actor
// contract grows only at architecture revision, never per application.
//
// THE LAST ROW IS NEW, AND IT IS THE ONE THAT SURPRISES. `agent-run-7f` appears
// twice with opposite verdicts: refused when the run itself raised the Request,
// granted when the SPINE did (`spine:consumer`, the consumer's own principal for
// conflations, faults, blown deadlines and the drain seal). Nothing about the
// confirmer changed — the REQUESTER did, and this gate compares principals, so a
// drain-requested seal is no longer a self-confirm. Pinned in both ports by
// `14.3 — the drain-requested seal and its confirmer`.
//
// A refusal is COMMITTED as `ToolResult.Refused`, so a re-fold reproduces the
// verdict without calling the authorization seam again — G9 satisfied, and G6's
// and G6's "captured as an ordered G9 fixture" are the same mechanism, not two.

import type { Authorization } from "../ports/authorization";
import type { Signature } from "../pure/actor";
import type { SealedResult } from "../pure/tool-result";
import { isSpineResult, refused, seal } from "../pure/tool-result";
import type { Registry } from "./action";

export function gate<S>(
  result: SealedResult,
  sig: Signature,
  state: S,
  registry: Registry<S>,
  authz: Authorization<S>,
): SealedResult {
  // Already a spine verdict (no verb ran, or an input failed to decode) —
  // there is nothing to gate, and it is already the thing that gets committed.
  if (isSpineResult(result)) return result;
  const verb = registry.get(result.tool);
  // Unreachable via resolveAction; kept total rather than thrown.
  if (verb === undefined) return seal(refused(result.tool, "no registered verb"));
  switch (verb.kind) {
    case "Reversible":
      return result;
    case "Irreversible": {
      const requester = verb.requestedBy(state, result);
      if (requester === null) return seal(refused(result.tool, "no pending request"));
      if (requester === sig.authority) {
        return seal(
          refused(
            result.tool,
            "self-confirm: the confirming authority is the requesting authority",
          ),
        );
      }
      return authz.mayConfirm(sig, result, state)
        ? result
        : seal(refused(result.tool, "authority may not confirm this action"));
    }
    default: {
      const _never: never = verb;
      return _never;
    }
  }
}
